### PR TITLE
[WIP][Typing/typecore.ml]: Added warning 68, matching a function type

### PR DESCRIPTION
**Pull request by muskangarg21**
#6881
@Octachron 
- This test case seems to be the problem now.
<pre>
# match fun x -> x with x -> x, x;;  
Warning 68: Trying to match a function type.
- : ('a -> 'a) * ('b -> 'b) = (<fun>, <fun>)
</pre>

- For now I have added a very basic test case, will add more after your suggestions.
- The descriptions and name associated to the newly added warning 68 needs to be discussed.
I will improve the code structure and add the Changes files once the PR reaches to its working state.

Thanks a lot for all the help

### CODE REVIEW

File: utils/warnings.ml [hunk 1]
```diff
@@ -775,6 +779,8 @@ let descriptions =
    64, "-unsafe used with a preprocessor returning a syntax tree";
    65, "Type declaration defining a new '()' constructor";
    66, "Unused open! statement";
+   (* 67, "ask octachron" *)
```
Comment (Octachron) [inline utils/warnings.ml:782]: Indeed, the short description for the `unused functor argument` warning is missing. If you want to add the missing text, it would make for a fine separate PR. Otherwise, myself (or @trefis) will fix the issue.

Comment (muskangarg21) [reply utils/warnings.ml:782]: Okay I ll add it.

**Comment by Octachron** [comment]
Concerning the failing tests, you can deactivate the warning locally with `match[@warning "-68"] `

File: typing/typecore.ml [hunk 2]
```diff
@@ -2697,11 +2705,23 @@ and type_expect_
...
+       end_def ();
       if maybe_expansive arg then lower_contravariant env arg.exp_type;
       generalize arg.exp_type;
       let cases, partial =
         type_cases Computation env arg.exp_type ty_expected true loc caselist in
+        let contains_exception_pattern pat =
+        exists_general_pattern { f = fun (type k) (p : k general_pattern) ->
+          match p.pat_desc with
+          | Tpat_exception _ -> true
+          | _ -> false } pat
```
Comment (Octachron) [inline typing/typecore.ml:2717]: This is a quite self-contained function that is used in two different places. It is often better in this situation to make the function a toplevel helper function shared between the two call sites.

**Comment by trefis** [comment]
Sorry to be the grumpy one, but I think this PR is premature. I don't think there's a consensus around the idea behind this PR, i.e. I don't think the silence that followed @damiendoligez's proposal was an approving one.

I am personally not convinced by the approach, matching on a function type is not *necessarily* a mistake. The issue lies with the ignoring of the result, not with the match itself. And while we do not have a satisfying answer to that issue yet, I'm not keen on adding "workaround" warnings.

**Comment by Octachron** [comment]
You are perfectly right to express your disagreement, and thanks for stating it early.

Do you have in mind a precise example in which matching on a syntactic value that is function (or any other type that cannot be destructured) is not a mistake?

**Comment by Gbury** [comment]
Matching on a partial application of a function to catch an exception in the construction of a cloture might be a reasonable use I'd guess ?

**Comment by Octachron** [comment]
I agree, this is why I added the "syntactic value" clause. And the current code only emits a warning if there are not exception pattern in the cases of the match.

**Comment by Octachron** [comment]
I am closing this PR as non-consensual, partially subsumed by #10328 and not being currently active.
